pbj-core/pbj-grpc-helidon/docs/design.md

```diff
@@ -0,0 +1,455 @@
+using the same thread. Here you can see that in the method call (`sayHelloStreamBidi`), we receive a `Flow.Subscriber`.
+Using the `onNext`, `onError`, and `onComplete` methods of this subscriber we can send messages back to the client,
+terminate the connection due to errors, or close the connection gracefully. We also return a new `Flow.Subscriber` that
+will receive messages from the client. The PBJ-Helidon module will push messages from the client to this subscriber
+using the `onNext` method.
+
+```java
+public class GreeterServiceImpl implements GreeterService {
+    @Override
+    @NonNull
+    public Flow.Subscriber<HelloReply> sayHelloStreamBidi(@NonNull final Flow.Subscriber<HelloRequest> replies) {
+        return new Flow.Subscriber<>() {
+            @Override
+            public void onSubscribe(Flow.Subscription subscription) {
+                subscription.request(Long.MAX_VALUE); // turn off flow control if you want to
+            }
+
+            @Override
+            public void onNext(HelloRequest item) {
+                replies.onNext(HelloReply.newBuilder()
+                        .setMessage("Hello " + item.getName())
+                        .build());
+            }
+
+            @Override
+            public void onError(Throwable throwable) {
+                replies.onError(throwable);
+            }
+
+            @Override
+            public void onComplete() {
+                replies.onComplete();
+            }
+        };
+    }
+}
+```
+
+The bidirectional streaming pipeline is similar to the client streaming pipeline, except that the `replies` subscriber
+may be called more than once.
+
+![Bidirectional Streaming Pipeline](images/bidi-streaming.png)
+
+# Test Plan
+
+## Functional Tests
+
+1. `Greeter.sayHello` can be successfully called
+2. `Greeter.sayHelloStreamReply` can be successfully called
+3. `Greeter.sayHelloStreamRequest` can be successfully called
+4. `Greeter.sayHelloStreamBidi` can be successfully called
+5. If `Greeter.sayHello` is called with invalid input, a `GrpcException` is thrown and turns into the expected gRPC error on the client
+6. If `Greeter.sayHelloStreamReply` is called with invalid input, a `GrpcException` is thrown and turns into the expected gRPC error on the client
+7. If `Greeter.sayHelloStreamReply` is called and sends `onError`, the client receives an error and the connection is terminated
+8. If `Greeter.sayHelloStreamRequest` is called with invalid input, a `GrpcException` is thrown and turns into the expected gRPC error on the client
+9. If `Greeter.sayHelloStreamRequest` is called and sends `onError`, the client receives an error and the connection is terminated
+10. If `Greeter.sayHelloStreamBidi` is called with invalid input, a `GrpcException` is thrown and turns into the expected gRPC error on the client
+11. If `Greeter.sayHelloStreamBidi` is called and sends `onError`, the client receives an error and the connection is terminated
```

Comment:
It may not be required to terminate the connection in this case.
Example:
Streaming validation server receives a stream of data items, and either acknowledges or errors for each one.
If an item is ack'd, that item is valid
If an item is error'd, that item is invalid
If the server sends "complete", then the server is terminating the connection (perhaps due to maximum single-stream duration, server shutdown, etc...)

Reply:
In the gRPC spec, the only way to send a gRPC error is by sending the trailers with endStream. The application can create their own messages that represent application level errors, rather than gRPC errors, and it would work as you describe. But every onError results in a gRPC error.

Reply:
Interesting, I did not know that.
That seems like something we may wish to detail as a limitation of the module and consider carefully in server design. Having connections terminate for an application error can be far more disruptive than we'd prefer, depending on situation.